pkg/reconciler/cache/publishedresources/publishedresources_reconcile_replication.go

```diff
@@ -0,0 +1,147 @@
+	dynamicCacheClient             kcpdynamic.ClusterInterface
+	kcpCacheClient                 kcpclientset.ClusterInterface
+	cacheKcpInformers              kcpinformers.SharedInformerFactory
+	discoveringDynamicKcpInformers *informer.DiscoveringDynamicSharedInformerFactory
+	callback                       func(obj interface{})
+	controllerRegistry             *controllerRegistry
+}
+
+func (r *replication) reconcile(ctx context.Context, publishedResource *cachev1alpha1.PublishedResource) (reconcileStatus, error) {
+	logger := klog.FromContext(ctx)
+	logger.Info("reconciling published resource", "publishedResource", publishedResource.Name)
+
+	gvr := schema.GroupVersionResource{
+		Group:    publishedResource.Spec.Group,
+		Version:  publishedResource.Spec.Version,
+		Resource: publishedResource.Spec.Resource,
+	}
+	cluster := logicalcluster.From(publishedResource)
+
+	clusterName := logicalcluster.From(publishedResource)
+	controllerName := fmt.Sprintf("%s.%s.%s.%s.%s", clusterName, gvr.Version, gvr.Resource, gvr.Group, publishedResource.Name)
+	// TODO: Add locking here when multiple workers are supported.
+	controller := r.controllerRegistry.get(controllerName)
+	// We setup controller even if we are deleting. This is to ensure that we can purge the cache.
+	// If for some reason was dead, we will recreate it.
+	if controller == nil {
+		// Global informer is based on the CachedResource type and we construct index based on the schema labels.
+		controllerCtx, cancel := context.WithCancel(ctx)
+
+		global := r.cacheKcpInformers.Cache().V1alpha1().PublishedObjects()
+
+		// Local informer is based on the specific types we want to replicate.
+		// TODO: use ClusterWithContext
+		local, err := r.discoveringDynamicKcpInformers.Cluster(cluster).ForResource(gvr)
+		if err != nil {
+			logger.Error(err, "Failed to get local informer for resource", "resource", gvr)
+			cancel()
+			return reconcileStatusStopAndRequeue, err
+		}
+		// TODO(mjudeikis): dynamic RestMapper!
+		replicated := &replicationcontroller.ReplicatedGVR{
+			Kind:   strings.TrimSuffix(strings.Title(gvr.Resource), "s"), //nolint:staticcheck
+			Local:  local.Informer(),
+			Global: global.Informer(),
+		}
+		replicationcontroller.InstallIndexers(replicated)
+		callback := func() {
+			r.callback(publishedResource)
+		}
+
+		c, err := replicationcontroller.NewController(
+			r.shardName,
+			r.dynamicCacheClient,
+			r.kcpCacheClient,
+			gvr,
+			replicated,
+			callback,
+			labels.SelectorFromSet(publishedResource.Spec.LabelSelector.MatchLabels),
+		)
+		if err != nil {
+			cancel()
+			return reconcileStatusContinue, err
+		}
+
+		go replicated.Local.Run(ctx.Done())
+		go replicated.Global.Run(ctx.Done())
+
+		if !cache.WaitForCacheSync(ctx.Done(), replicated.Local.HasSynced, replicated.Global.HasSynced) {
+			cancel()
+			return reconcileStatusContinue, fmt.Errorf("failed to wait for informers to sync")
+		}
+
+		go func() {
+			c.Start(controllerCtx, 1)
+			cancel()
```

Comment:
```suggestion
```

This can be omitted, if `.Start` finishes the context has already been cancelled.